repose-aggregator/components/services/atom-feed-service/atom-feed-service-impl/src/main/scala/org/openrepose/nodeservice/atomfeed/impl/auth/OpenStackIdentityV2AuthenticatedRequestFactory.scala

```diff
@@ -67,59 +70,75 @@ class OpenStackIdentityV2AuthenticatedRequestFactory(configuration: OpenStackIde
+      s"$serviceUri$TOKENS_ENDPOINT",
+      Map(CommonHttpHeader.ACCEPT.toString -> MediaType.APPLICATION_JSON,
+        CommonHttpHeader.TRACE_GUID.toString -> tracingHeader).asJava,
+      Json.stringify(Json.obj(
+        "auth" -> Json.obj(
+          "passwordCredentials" -> Json.obj(
+            "username" -> username,
+            "password" -> password)))),
+      MediaType.APPLICATION_JSON_TYPE
+    ))
+
+    akkaResponse match {
+      case Success(serviceClientResponse) if Option(serviceClientResponse).isDefined =>
+        serviceClientResponse.getStatus match {
+          case statusCode@(SC_OK | SC_NON_AUTHORITATIVE_INFORMATION) =>
+            val jsonResponse = Source.fromInputStream(serviceClientResponse.getData).getLines().mkString("")
+            Try(Json.parse(jsonResponse)) match {
+              case Success(json) =>
+                Try(Success((json \ "access" \ "token" \ "id").as[String])) match {
+                  case Success(token) =>
+                    logger.debug("Successfully fetched and parsed token from identity service")
+                    cachedToken = Some(token.get)
+                    token
+                  case Failure(f) =>
+                    logger.error(s"Response from the identity service was not in JSON format as expected: ${f.getLocalizedMessage}")
+                    Failure(new Exception("Response from the identity service was not in JSON format as expected", f))
+                }
+              case Failure(f) =>
+                logger.error(s"Response from the identity service was not in JSON format as expected: ${f.getLocalizedMessage}")
+                Failure(new Exception("Response from the identity service was not in JSON format as expected", f))
+            }
+          case statusCode =>
+            logger.error(s"Failed to retrieve token from the identity service, status code: $statusCode")
+            Failure(new Exception(s"Failed to retrieve token from the identity service, status code: $statusCode"))
+          case _ =>
+            logger.error("Failed to retrieve token from the identity service.")
+            Failure(new Exception("Failed to retrieve token from the identity service."))
+        }
+      case Failure(f) =>
+        logger.error(s"Failed to retrieve token from the identity service due to: ${f.getLocalizedMessage}")
+        Failure(new Exception("Failed to retrieve token from the identity service.", f))
+      case _ =>
+        logger.error("Failed to retrieve token from the identity service.")
+        Failure(new Exception("Failed to retrieve token from the identity service."))
     }
   }
 
-  private def parseTokenFromJson(tokenStream: InputStream): String = {
-    val contentString = Source.fromInputStream(tokenStream)(Codec.UTF8).mkString
+  override def onInvalidCredentials(): Unit = cachedToken = None
 
-    (Json.parse(contentString) \ "access" \ "token" \ "id").as[String]
+  def setConnectionPoolId(poolId: String): Unit = {
```

Comment:
It seems that it's not needed, but is the behavior of the app right since it's not being called?